datatypes/src/main/java/org/hyperledger/besu/datatypes/BLSPublicKey.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright contributors to Hyperledger Besu
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.datatypes;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import org.hyperledger.besu.ethereum.rlp.RLPException;
+import org.hyperledger.besu.ethereum.rlp.RLPInput;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.DelegatingBytes;
+
+/** A BLS public key. */
+public class BLSPublicKey extends DelegatingBytes
+    implements org.hyperledger.besu.plugin.data.PublicKey {
```

Comment:
Yea I was debating whether to put it under `datatypes` or `crypto`. My line of thought is `crypto` is more dedicated to classes that need crypto-related operations such as encryption and such. `BLSPublicKey` is mainly a CL thing, all we are doing in EL is just view this as a 48-byte field and we don't perform any crypto operation for `BLSPublicKey` hence putting it under `datatypes`. I am happy to move it to `crypto` if you think otherwise.

Reply:
The datatypes package does seem like a more appropriate place than crypto as it is more focused on operations.

However, the intent of the datatype module is for broad-based types. So I don't think the `DepositWithdrawalCredential` would qualify and should probably just be a Bytes32 since we only care about bytes anyway or move it closer to where it is needed rather than in the top-level module.

Reply:
> The datatypes package does seem like a more appropriate place than crypto as it is more focused on operations.
> 
> However, the intent of the datatype module is for broad-based types. So I don't think the `DepositWithdrawalCredential` would qualify and should probably just be a Bytes32 since we only care about bytes anyway or move it closer to where it is needed rather than in the top-level module.

I agree we should just remove `DepositWithdrawalCredential` and make the field as Bytes32. What do you think about `BLSPublicKey` and `BLSSignature`? AFAIK we are not doing anything special for these 3 fields as we only care about the bytes. Having dedicated data classes are kind of like syntactic sugar

Reply:
I think BLSPublicKey and BLSSignature are OK as new types since they are quite a general concept and could imagine them being reused in the future. 
Whereas DepositWithdrawalsCredential is quite specific Bytes32 is better IMO...it's not clear to me from the EIP what format this credential takes exactly, maybe there's an existing data type that is appropriate?

Possibly BLSPublicKey and BLSSignature shouldn't live in data types though since they are not broadly used. Maybe should live alongside Deposit.java for now?

Reply:
> I think BLSPublicKey and BLSSignature are OK as new types since they are quite a general concept and could imagine them being reused in the future. Whereas DepositWithdrawalsCredential is quite specific Bytes32 is better IMO...it's not clear to me from the EIP what format this credential takes exactly, maybe there's an existing data type that is appropriate?
The format of withdrawal credential is one of its kind and is rather specific. To summarize what it contains: there are two types one with 0x00 and one with 0x01 prefix. The first type is a hash of public key with the first byte being replaced by 0. The second type is simply a 20-byte eth1 address. I think there are some discussions in CL on coming up with new prefixes and new encoding for the withdrawal credential. So in short, I don't think we can use existing data type for it.
> 
> Possibly BLSPublicKey and BLSSignature shouldn't live in data types though since they are not broadly used. Maybe should live alongside Deposit.java for now?
Yea I agree. We can move them to Deposit.